cmd/kubeadm/app/phases/controlplane/manifests.go

```diff
@@ -109,6 +119,12 @@ func CreateStaticPodFiles(manifestDir, patchesDir string, cfg *kubeadmapi.Cluste
 			klog.V(2).Infof("[control-plane] adding volume %q for component %q", v.Name, componentName)
 		}
 
+		if usersAndGroups != nil {
```

Comment:
usersAndGroups cannot be nil unless we returned an error earlier.

it's actually more accurate to have:
`if features.Enabled(cfg.FeatureGates, features.RootlessControlPlane) && !isDryRun {` here.

you can also move features.Enabled(cfg.FeatureGates, features.RootlessControlPlane) to a local constant.

the utils we added in user_linux.go have no dry run flag.
adding it seems a bit too verbose.

instead we should do the following here and above:

```
if isDryRun {
    // print message that we "would" add user and groups
} else {
   if runAsRoot {
       // do foo
   }
}
```

Reply:
> usersAndGroups cannot be nil unless we returned an error earlier.
It can be if someone on !linux enables the feature gate.

Reply:
sorry I did not understand that. Can you elaborate a bit?

Reply:
line 102 must be:
```
if isDryRun {
    // print message that we "would" add user and groups
    // there are dry run formatting examples in the code base
} else {
   if runAsRoot {
       // do foo
   }
}
```

Reply:
can you point me to a code location where we do dry-run message printing?

Reply:
Done.